src/ImageSharp/Formats/Png/Zlib/DeflaterHuffman.cs

```diff
@@ -40,8 +40,6 @@ internal sealed unsafe class DeflaterHuffman : IDisposable
         // probability, to avoid transmitting the lengths for unused bit length codes.
         private static readonly int[] BitLengthOrder = { 16, 17, 18, 0, 8, 7, 9, 6, 10, 5, 11, 4, 12, 3, 13, 2, 14, 1, 15 };
```

Comment:
We might well be better off using the `ReadOnlySpan<byte>` here also and casting. 

I know we do that in the jpeg encoder for `uint` and `BitCountLut` with, as I recall, a net gain in performance.

Reply:
Done in https://github.com/SixLabors/ImageSharp/pull/1133/commits/5b7ac75dfa637f40763b75bcd378205a79394e5d.
Should I also add some bounds check here, or are all those accesses guaranteed to be safe?

Reply:
I couldn't tell you for certain I'm afraid. That port was an act of desperation over understanding.

Reply:
I see. Should I just switch those accesses to a direct `Span<T>[int]` access then? 🤔

Reply:
If we lack understanding, we shall go for safety.

Btw if we use the data in `Bit4Reverse` for indexing, elements shouldn't be `byte`-s. It's more work for the CPU:
https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcAMACbckB0ASgK4B2wAllAKYEDCA9lAA6UA2NATgMrcBulAMY0IAbnTo4AZjwJs9bAG90uXKrUsulfmGA08SAGzYARgE992AJJkAJjQAet4NIQAKLjQBmZywZYAGmxKChD7JwBKDVwVNDUEvAB2bABVMggwbzoAQTs7Tx9sIPCHR0iJeLUAXxjsOq0dPQN8EwsrWzKAIX9C33aA4IHSqLq4xLU4FPTM7II8gq9fEtCyirratGqgA==

Reply:
@antonfirsov not sure what you mean there 🤔

We can't use the `static ReadOnlySpan<T>` trick with `int` values, as those are not supported. By that I mean, you'd get a new array being allocated every time, that's because `byte` is not endianness-aware, while `int` is not. As for the actual value being used for offsetting, I checked all the various `Unsafe.Add` calls, and the offsetting index is always passed as an `int` value, not as a `byte`. Could you clarify what you meant there? I'm not following 😅

Reply:
Not an ASM expert, so the terminology might be incorrect, but looks something like this:
Byte can't be directly used to address memory or added to a memory address, therefore it has to be loaded into another register first, thus the extra `movzx eax, dl` in my gist.

A possible workaround is to create `ReadOnlySpan<byte>` of `*4` size with little padding, and unsafe cast it, but then you have to deal with endianness, which will be super-complicated. An other solution is to not use the `.text` block, for these blobs since, you only save 1 `memcpy` at initialization time with `.text`, while indexing will be slower each and every time invoking the methods.

Reply:
I get what you're saying now, yeah 👍
I'm not sure I understand where you're seeing that with `Bit4Reverse` though.
Here's where that is used:
https://github.com/SixLabors/ImageSharp/blob/4b0dfd11c0267905ea8f004833db2618bf4b2ab7/src/ImageSharp/Formats/Png/Zlib/DeflaterHuffman.cs#L384-L392

All the indexing values there are `int` already 🤔
I mean, those are actually the same indices that were being used before as well. I'm not sure I'm following here, where is the `byte` offsetting peformed?

Reply:
Ahh sorry, I was wrong the values for `Bit4Reverse` are not used for indexing into data. This is the data that gets indexed.

We should deal with Guard however to make sure we don't regress. Do you plan to fix it globally in SixLabors/SharedInfrastructure, or will you replace it in this PR with a manual check + ThrowHelper calls?

Reply:
Ahahah no problem, glad we could figure this out, I was wondering what was I missing there 😄

As for `Guard`, I'd say that's an unrelated improvement, since it's not something that will impact the functionality in any way, it's just a matter of further optimizations. I guess the best thing would be to just make another PR and fix the entire `Guard` class, with all the included APIs. Sounds good?

Reply:
If you can do it, that would be the best!

Reply:
Sure, always happy to bring more of my discoveries and optimizations to ImageSharp! 😄